shared/utils/lazy.hpp

```diff
@@ -0,0 +1,74 @@
+#include <atomic>
+#include <cstdint>
+#include <functional>
+#include <optional>
+
+/// @brief Thread-safe container which ensures its value is initialized only once.
+/// @tparam T Type of the contained value.
+template <typename T>
+class Once {
+    std::optional<T> value;
+    std::atomic_uintptr_t state_and_queue;
+
+    void initialize(std::function<T()> const& initializer);
+
+  public:
+    /// @brief Constructs an uninitialized `Once<T>`.
+    constexpr Once() : value(), state_and_queue(0) {}
+
+    Once(Once<T> const&) = delete;
+
+    /// @brief Gets the current value, if it is initialized.
+    std::optional<T const&> get() const {
+        return value;
+    }
+    //// @brief Gets the current value, if it is initialized.
+    std::optional<T&> get() {
+        return value;
+    }
+
+    /// @brief Gets the current value, or initializes it.
+    /// @param initializer Function to use for initialization if required.
+    T& get_or_initialize(std::function<T()> const& initializer) {
+        initialize(initializer);
+        return *value;
+    }
+};
+
+/// @brief Thread-safe, lazy-initialized value.
+/// @tparam T Type of the value.
+/// @tparam F Type of the initializer.
+template <typename T, typename F = std::function<T()>>
+class Lazy {
+    mutable Once<T> inner;
+    F initializer;
+
+  public:
+    /// @brief Construces an uninitialized `Lazy<T>`.
+    /// @param initializer_ Function to use for initialization on the first access.
+    Lazy(F const& initializer_) : inner(), initializer(initializer_) {}
+
+    operator T() const {
```

Comment:
Personally, I'm not sure how I feel about implicit conversion operators, but this is ultimately a non-issue The repeated code here bothers me a little, but making a macro for it is probably unnecessary.